### PR TITLE
Added option to show original locations and path in error

### PR DESCRIPTION
**Pull request by scf4**

### CODE REVIEW

**Comment by thebigredgeek** [comment]
@scf4 thanks!  Can you update tests?

**Comment by thebigredgeek** [comment]
Also, this might work better via an options hash versus two separate booleans.

**Comment by thebigredgeek** [comment]
Looks like it's failing because `undefined` won't serialize, it will simply be omitted.  You will probably need to update the code :(

**Comment by scf4** [comment]
Should be good now :)

**Comment by thebigredgeek** [comment]
Can you update the test to make sure that the error retains the location and path after being passed through formatError?  Also, can you update the docs?  Thanks for the work on this @scf4.

File: src/index.js [hunk 1]
```diff
@@ -11,53 +11,65 @@ class ApolloError extends ExtendableError {
...
         toString: () => JSON.stringify(d)
-      })
+      }),
     ]));
 
     this._name = name;
     this._humanized_message = message || '';
     this._time_thrown = t;
     this._data = d;
+    this._locations = (opts.showLocations && arguments[2] && arguments[2].locations)
```
Comment (thebigredgeek) [inline src/index.js]: Why are you using arguments here rather than opts?

File: test/spec.js [hunk 2]
```diff
@@ -8,15 +8,23 @@ describe('createError', () => {
...
 
     const e = new FooError({
       data: {
         foo: 'bar'
-      }
+      },
+      options: {
+        showLocations: true,
+        showPath: false,
+      },
```
Comment (thebigredgeek) [inline test/spec.js]: Please test formatError to make sure things are serializing properly

File: src/index.js (hunk 1 above)
Comment (scf4) [inline src/index.js]: It's a truthy check of the opts. The line was initially below `const t = (arguments[2] && arguments[2]...` and I was keeping it brief, but I'll make the code a little clearer tomorrow!

The arguments come from the originalError in formatError.

**Comment by scf4** [comment]
The error comes from the originalError object which is passed to formatError. What wouldn't be serialized properly?

My use case is preserving the path info so I know which query caused which error when running multiple. I don't personally need locations but it could be relevant to some!

**Comment by thebigredgeek** [comment]
It "should", but we want to add tests to make sure that if we create an error with such options, the options are preserved when the error is serialized out to the client

**Comment by thebigredgeek** [comment]
Will merge after tests are adding.  Wanna be super safe.  Also, still wondering the use case here.  Wouldn't I need to manually pass the location and path into the error data payload? If that's the case, are the booleans even necessary?

**Comment by thebigredgeek** [comment]
@scf4 bump :)

**Comment by scf4** [comment]
@thebigredgeek Apollo adds the path and locations to errors by default.

**Comment by thebigredgeek** [comment]
It would need to know to add them to the data object in the error constructor though, wouldn't it?

**Comment by scf4** [comment]
formatError has originalError.path and originalError.locations. It passes them to the new custom error, which checks its opts (added in createError) to see if they should be returned in the object from serialize(). They don't get serialized into the `blah/::/blah` string because they don't exist at that point.

**Comment by thebigredgeek** [comment]
Ahh I see now.  Clever.  Nice!  Will merge in just a few

**Comment by scf4** [comment]
Awesome thanks!

**Pull request merged**